File: 2016/Robot2016/src/com/kennedyrobotics/commands/drive/TurnSpecifiedDegreesCommand.java
package com.kennedyrobotics.commands.drive;

import com.kennedyrobotics.Robot;

import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.command.Command;

/**
 *
 */
public class TurnSpecifiedDegreesCommand extends Command {

	private static final double DEGREES_PRECISION = 3;
	private final Timer timer = new Timer();
	
	private boolean finished = false;
	private double goalAngle = 0.0;
	private final double degreesToTurn;
	
    public TurnSpecifiedDegreesCommand(double degreesToTurn) {
    	if(Math.abs(degreesToTurn) > 180) {
    		throw new IllegalArgumentException("Degrees not in range +-180");
    	}
    	requires(Robot.chassis);
    	this.degreesToTurn = degreesToTurn;
    }

    // Called just before this Command runs the first time
    protected void initialize() {
    	finished = false;
    	
    	Robot.sensorSubsystem.getNavX().zeroYaw();
    	System.out.println(Robot.sensorSubsystem.getYaw());
    	goalAngle = degreesToTurn;
    	
    	timer.reset();
    	timer.start();
    	
    	System.out.println("[TurnSpecifiedDegreesCommand] Going to turn " + degreesToTurn);
    }

    // Called repeatedly when this Command is scheduled to run
    protected void execute() {
    	if(!timer.hasPeriodPassed(0.2)) {
    		return;
    	}
    	
    	double currentAngle = Robot.sensorSubsystem.getYaw();
    	double angleDifference = goalAngle - currentAngle;
    	//angleDifference %= 180;
    	System.out.println("Angle Difference " + angleDifference + " gaolAngle " + goalAngle + " currentAngle " + currentAngle);
    	
    	double turnRate = 0;
    	if(Math.abs(angleDifference) <  DEGREES_PRECISION) {
        	Robot.chassis.stop();
    		finished = true;
    		return;
    	} else {
    		if(Math.abs(angleDifference) > 20) {
    			turnRate = 0.9;
    		} if(Math.abs(angleDifference) > 10) {
    			turnRate = 0.8;
    		} else { 
    		
    			turnRate = 0.7;
    		}
    	}
    	
    	if(angleDifference > 0) {
    		turnRate = -turnRate;
    	}
    	
    	Robot.chassis.arcadeDrive(0.0, turnRate);
    	
    }

    // Make this return true when this Command no longer needs to run execute()
    protected boolean isFinished() {
        return finished;
    }

    // Called once after isFinished returns true
    protected void end() {
    	System.out.println("[TurnSpecifiedDegreesCommand] completed.");
    	Robot.chassis.stop();
    	timer.stop();
    }

    // Called when another command which requires one or more of the same
    // subsystems is scheduled to run
    protected void interrupted() {
    	System.out.println("[TurnSpecifiedDegreesCommand] Interrupted.");
    	Robot.chassis.stop();
    	timer.stop();
    }
}

File: 2017/2017Robot/src/Commands/MotionProfiling/TurnSpecifiedDegreesCommand.cpp
#include "TurnSpecifiedDegreesCommand.h"
#include "math.h"

// If we're within this # of degrees to target, then we're good.

#define DEGREES_PRECISION 1.5


/*
 * This command turns the number of degrees specified in the dashboard field.
 */
//Positive angles to the right
TurnSpecifiedDegreesCommand::TurnSpecifiedDegreesCommand(double degreesToTurn)
	: CommandBase("TurnSpecifiedDegreesCommand") {
	Requires(chassis.get());
	this->goalAngle = degreesToTurn;
	this->timer = new Timer();
	this->finished = false;
}

// Called just before this Command runs the first time
void TurnSpecifiedDegreesCommand::Initialize() {
	this->finished = false;
	chassis->EnableBrake();
//	chassis->SetReversed(false);
	sensorSubsystem->ZeroYaw();
	this->timer->Reset();
	this->timer->Start();
	printf("TurnSpecifiedDegreesCommand %f\n", this->goalAngle);
}

// Called repeatedly when this Command is scheduled to run
void TurnSpecifiedDegreesCommand::Execute() {
//	double currentAngle = sensorSubsystem->GetIMU()->GetYaw();
	double currentAngle = sensorSubsystem->GetYaw();
	double angleDifference = goalAngle - currentAngle;
	
	SmartDashboard::PutNumber("Goal angle", goalAngle);
	SmartDashboard::PutNumber("Angle difference", fabs(angleDifference));
	printf("Angle difference %f goalAngle %f currentAngle %f\n", angleDifference, goalAngle, currentAngle);
	
	double turnRate = 0;
	
	if (fabs(angleDifference) < DEGREES_PRECISION) {
		chassis->Stop();
		finished = true;
	} else {
		// We slow our rate of turn as we get close to the angle we want.
		// These values are guesses.  A PID would be better here.

		if(fabs(angleDifference) > 60.0)  {
			turnRate = 0.30;
		} else if(fabs(angleDifference) > 45.0) {
			turnRate = 0.25;
		} else if(fabs(angleDifference) > 20.0) {
			turnRate = 0.20;
		} else if(fabs(angleDifference) > 15.0) {
			turnRate = 0.15;
		} else {
			turnRate = 0.13;
		}

		SmartDashboard::PutNumber("turn rate", turnRate);
		
		if (angleDifference < 0) {
			chassis->ArcadeDrive(0.0, -turnRate);
		} else {
			chassis->ArcadeDrive(0.0, turnRate);
		}
	}
}

// Make this return true when this Command no longer needs to run execute()
bool TurnSpecifiedDegreesCommand::IsFinished() {
	return finished || this->timer->HasPeriodPassed(5);
}

// Called once after isFinished returns true
void TurnSpecifiedDegreesCommand::End() {
	printf("TurnSpecifiedDegreesCommand completed.\n");
	this->cleanUp();
}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void TurnSpecifiedDegreesCommand::Interrupted() {
	printf("TurnSpecifiedDegreesCommand interrupted.\n");
	this->cleanUp();
}

void TurnSpecifiedDegreesCommand::cleanUp() {
	chassis->Stop();
//	chassis->DisableBrake();
	this->timer->Stop();
}
